test-data/unit/check-based-typeguard.test

```diff
@@ -23,3 +23,351 @@ a: object
+    reveal_type(o)  # N: Revealed type is "int"
+
+if (lambda _, y: guard(y))(o, 1):
+    reveal_type(o)  # N: Revealed type is "object"
+
+class A:
+    def guard(self) -> 'self' is "B": ...
+class B(A): ...
+a: A
+assert (lambda x: x.guard())(a)
+reveal_type(a)  # N: Revealed type is "__main__.B"
+[builtins fixtures/tuple.pyi]
+
+
+[case testTypeguardSubtype]
+def guard(x: int) -> 'x' is str:  # E: A type-guard's type must be assignable to its parameter's type. (guard has type "str", parameter has type "int")  [typeguard-subtype] \
+                                  # N: If this is correct, try making it an intersection with the parameter type
+    ...
+
+
+[case testOtherArg]
+def guard(x: object, y: object) -> 'y' is int: ...
+
+a: object
+b: object
+
+if guard(a, b):
+    reveal_type((a, b))  # N: Revealed type is "(object, int)"
+
+if guard(x=a, y=b):
+    reveal_type((a, b))  # N: Revealed type is "(object, int)"
+
+if guard(y=a, x=b):
+    reveal_type((a, b))  # N: Revealed type is "(int, object)"
+
+if guard(y=b, x=a):
+    reveal_type((a, b))  # N: Revealed type is "(object, int)"
+[builtins fixtures/tuple.pyi]
+
+
+[case testUnsupportedDoubleGuardIIFE]
+# unsupported scenario
+def guard(x: object) -> 'x' is int: return True
+
+a: object
+b: object
+
+if (lambda x, y: guard(x) and guard(y))(a, b):
+    # both of these should be "int"
+    reveal_type((a, b))  # N: Revealed type is "(object, object)"
+[builtins fixtures/tuple.pyi]
+
+
+[case testTargetDoesntExist]
+def guard(x: object) -> 'y' is int: ...  # E: Cannot find parameter "y"  [name-defined]
+
+
+[case testInvalidType]
+def guard(x: object) -> 'x' is Sus: ...  # E: Name "Sus" is not defined  [name-defined]
+
+
+[case testExplicitSelf]
+from typing_extensions import TypeGuard
+
+class A:
+    def guard(self, x: object) -> TypeGuard[int]: ...
+
+a: object
+assert A.guard(A(), a)
+reveal_type(a)  # N: Revealed type is "int"
+[builtins fixtures/tuple.pyi]
+
+
+[case testGuardSelf]
+class A:
+    def guard(self) -> 'self' is "B": ...
+
+class B(A): ...
+
+a: A
+assert a.guard()
+reveal_type(a)  # N: Revealed type is "__main__.B"
+
+
+[case testGuardCls]
+from typing import Type
+class A:
+    @classmethod
+    def guard(cls) -> 'cls' is Type["B"]: ...
+
+class B(A): ...
+
+a: Type[A]
+assert a.guard()
+reveal_type(a)  # N: Revealed type is "type[__main__.B]"
+[builtins fixtures/classmethod.pyi]
+
+
+[case testPositionalGuard]
+def guard(x: object, /) -> 'x' is int: ...
+
+a: object
+assert guard(a)
+reveal_type(a)  # N: Revealed type is "int"
+reveal_type(guard)  # N: Revealed type is "def (object) -> argument 1 is int"
+[builtins fixtures/tuple.pyi]
+
+
+[case testUnquotedFuture]
+from __future__ import annotations
+def guard(x: object) -> x is int: ...
+[builtins fixtures/tuple.pyi]
+
+
+[case testUnquotedPast]
+def guard(x: object) -> x is int:  # E: You need to put quotes the variable reference, the entire type-guard, or enable `__future__.annotations`  [misc]
```

Comment:
remove "You need to put quotes the variable reference" because it raises a warning at runtime